Add docs for async udfs (#72)
ArroyoSystems/arroyo-docs: sql/async_udfs.mdx

---
title: Asynchronous User-Defined Functions
description: 'Run Async UDFs'
---

Arroyo's UDFs can be asynchronous.

> This is useful for UDFs that need to make network requests or perform other asynchronous operations.

Reviewer: It's maybe worth emphasizing that the distinction is that normal UDFs need to run very quickly (<<1ms) because they block processing, while asynchronous UDFs can run in parallel and can therefore take much longer to execute while still maintaining good throughput.

I would also include the caveat that async UDFs cannot block—they must use async operations or use spawn_blocking (https://docs.rs/tokio/latest/tokio/task/fn.spawn_blocking.html)

Author: Yup, done.

---
title: Asynchronous User-Defined Functions
description: 'Run Async UDFs'
---

Arroyo's UDFs can be asynchronous.
This is useful for UDFs that need to make network requests or perform other asynchronous operations.
To define an asynchronous UDF, use the `async` keyword in the function definition.

## Options

Async UDFs have a few configuration options that can be defined in a special comment at the top of the UDF,
similar to dependencies, under a `[udfs]` header.

> | Option | Example | Description |

Reviewer 2: Might be cleaner to just have a `Default` column.

Author: The examples are the defaults so I just renamed the column

This can be used to store resources that should be shared across executions of the UDF.

The `Context` struct must implement a pre-defined trait called `UdfContext`.
The trait is defined as follows:

```rust
#[async_trait]
pub trait UdfContext: Sync {
    async fn init(&self) {}
    async fn close(&self) {}
}
```

In addition to the above trait, the `Context` struct must also define a `new` function (with no arguments).
The `new` and `init` functions will be called once when the job starts,
and the `close` function will be called once when the job ends.

If a `Context` is defined, an `Arc` reference to it will be passed as the first argument to the UDF.

## Example

This example defines an async UDF that's used to fetch a user's name from a database.

```rust
/*
[dependencies]
tokio = { version = "1", features = ["full"] }
tokio-postgres = "0.7"
async-trait = "0.1.68"
arroyo-types = { path = "../../../arroyo-types" }

[udfs]
async_results_ordered = true
async_max_concurrency = 10
async_timeout_seconds = 5
*/

use tokio_postgres::{NoTls, Error, Client};
use async_trait::async_trait;

> use arroyo_types::UdfContext;

Reviewer 2: Can we do something so users don't need to know about Arroyo packages they need to import? Alternately, just qualify the UdfContext in the implementation block.

Reviewer: Can we add the arroyo-types dependency automatically? The exact path is dependent on our build layout which is definitely going to be changing, and it feels like it should be an implementation detail that shouldn't be exposed to users.

Author: Ok I've made that dependency automatic in https://github.com/ArroyoSystems/arroyo/pull/487 and updated this example.

use std::sync::Arc;
use tokio::sync::Mutex;

pub struct Context {

>     client: Mutex<Option<Client>>,

Reviewer 2: Putting this on a Mutex means that only one call can have access to the client at any given time, which isn't great. Could this be done with a RWLock instead?

Author: Good call, done.

---
title: Asynchronous User-Defined Functions
description: 'Run Async UDFs'
---

Arroyo's UDFs can be asynchronous.
This is useful for UDFs that need to make network requests or perform other asynchronous operations,
and therefore should not block processing. Since asynchronous UDFs are executed in parallel,
we can maintain high throughput even with longer-running UDFs.

> Note that async UDFs should not block (i.e. `await`) as this will block the main processing thread.

Reviewer: This is phrased in a way that might be confusing for users.

In Rust async code, you must use async primitives (i..e, things you `await` on) rather than blocking code. I.e., `std::time::sleep(..)` is bad because it blocks, but `tokio::time::sleep(..).await` is good because it hooks into the async machinery to defer processing.

Instead, they should use asynchronous operations or [spawn_blocking](https://docs.rs/tokio/latest/tokio/task/fn.spawn_blocking.html).

To define an asynchronous UDF, use the `async` keyword in the function definition.

## Options

Async UDFs have a few configuration options that can be defined in a special comment at the top of the UDF,
similar to dependencies, under a `[udfs]` header.

| Option | Default | Description |
| --- | --- | --- |
| `async_results_ordered` | `true` | Whether to return results in the order they were requested (ordered) or as they are available (unordered). |
| `async_max_concurrency` | `100` | The maximum number of concurrent executions of this UDF. |

> | `async_timeout_seconds` | `10` | The maximum number of seconds to wait for a result before timing out. Arroyo currently drops the record if the UDF times out or panics. |

Reviewer 2: ```suggestion
| `async_timeout_seconds` | `10` | The maximum number of seconds to wait for a result before timing out. Arroyo currently retries if `async_results_ordered` is false, and panics if it is true. |
```

## Context

> Async UDFs can optionally define a `Context` struct that will be passed to each execution of the UDF.

Reviewer 2: Do we need to specify the magic naming convention for the UDF?

Author: The udf can be named anything, it's just the `Context` that must be named `Context`, and the `context` parameter. I've made that more clear below.
